File: 1_Algorithmic_Toolbox/week3_greedy_algorithms/3_car_fueling/car_fueling.cpp
#include <iostream>
#include <vector>

using std::cin;
using std::cout;
using std::vector;
using std::max;

int compute_min_refills(int dist, int tank, vector<int> & stops) {
    // write your code here

    if (tank >= dist) {
        return 0;
    }

    if (tank < stops[0]) {
        return -1;
    }

    int min_refills = 0;
    int prev_stop = 0;

    int gas_left_in_the_tank = tank;

    for(unsigned int i = 0; i < stops.size(); i++) {
        // cout << "Current Stop: " << stops[i] << std::endl;
        int distance_covered = stops[i] - prev_stop;
        gas_left_in_the_tank -= distance_covered;

        if (gas_left_in_the_tank < 0) {
            return -1;
        }

        if (gas_left_in_the_tank < (stops[i + 1] - stops[i])) {
            // cout << "Refuelling" << std::endl;
            min_refills += 1;
            gas_left_in_the_tank = tank; // Refuel
        }
        prev_stop = stops[i];
    }

    // Covering the left over distance to the destination from the final gas stop
    if (gas_left_in_the_tank < (dist - stops[stops.size() - 1])) {
        // cout << "Refuelling at the final gas stop" << std::endl;
        min_refills += 1;
        gas_left_in_the_tank = tank; // Refuel at the final stop
        gas_left_in_the_tank -= (dist - stops[stops.size() - 1]);
    }

    if (gas_left_in_the_tank < 0) {
        return -1;
    }

    return min_refills;
}


int main() {
    int d = 0;
    cin >> d;
    int m = 0;
    cin >> m;
    int n = 0;
    cin >> n;

    vector<int> stops(n);
    for (int i = 0; i < n; ++i)
        cin >> stops.at(i);

    cout << compute_min_refills(d, m, stops) << "\n";

    return 0;
}

File: 1_Algorithmic_Toolbox/week3_greedy_algorithms/3_car_fueling/car_fueling.py
# python3
import sys


def compute_min_refills(distance, tank, stops):
    # write your code here
    if (tank >= distance):
        return 0

    if (tank < stops[0]):
        return -1

    min_refills = 0
    prev_stop = 0

    gas_left_in_tank = tank

    for i in range(len(stops)):
        distance_covered = stops[i] - prev_stop
        gas_left_in_tank -= distance_covered

        # Ran out of the gas before reaching the next gas station
        if gas_left_in_tank < 0:
            return -1

        # Refill the gas if the car cannot reach the next gas station from the current
        if i < (len(stops) - 1) and (gas_left_in_tank < (stops[i+1] - stops[i])):
            gas_left_in_tank = tank # Refuel the tank
            min_refills += 1
        prev_stop = stops[i]

    # Checking if car needs to refuel at last gas station to reach destination
    if (gas_left_in_tank < (distance - stops[-1])):
        gas_left_in_tank = tank
        min_refills += 1
        gas_left_in_tank -= (distance - stops[-1])

    if gas_left_in_tank < 0:
        return -1

    return min_refills

if __name__ == '__main__':
    d, m, _, *stops = map(int, sys.stdin.read().split())
    print(compute_min_refills(d, m, stops))
